### DISCUSSION TITLE
media element attributes too verbose

### DISCUSSION THREAD

**Discussion by luwes** [muxinc/media-chrome#256] [Proposals]
Some media element attribute names seem very verbose and go a little against the way one would write standalone custom elements which was one of the goals of Media Chrome.

For example 
https://media-chrome-git-fork-luwes-improve-time-range-mux.vercel.app/examples/control-elements/media-thumbnail-preview.html

```html
<media-thumbnail-preview 
  media-preview-image="https://image.mux.com/O6LdRc0112FEJXH00bGsN9Q31yu5EIVHTgjTKRkKtEq1k/storyboard.jpg" 
  media-preview-coords="284 640 284 160"></media-thumbnail-preview>
```

I see that this is for partly done for name collision purposes in the propagation from the controller but wondering if there is a way around this?

This would be a lot neater and would still be the same clear to me

```html
<media-thumbnail-preview 
  src="https://image.mux.com/O6LdRc0112FEJXH00bGsN9Q31yu5EIVHTgjTKRkKtEq1k/storyboard.jpg" 
  coords="284 640 284 160"></media-thumbnail-preview>
```

---

Btw I think we might have to reverse the naming there?
`media-thumbnail-preview` -> `media-preview-thumbnail`

with this new element 

`media-preview-time-display`

**Comment by cjpillsbury** [comment]
I'm very much in favor of this (though it would be a major change). In fact, for pragmatic/forward reasons I'd favor removing the dash as well, e.g.

`media-has-played` becomes `hasplayed`

A major motivating factor are the "props" vs. "attribute" translations from frameworks, particularly React's plan, described here https://docs.google.com/document/d/1wwL_YZ5TnYorEEr2Lkeh9rgbnsynDXXI7tCejVcpE6I/edit#

**Reply by heff** [reply]
> particularly React's plan

Mind summarizing the relevant bits of React's plan?

**Reply by cjpillsbury** [reply]
For sure. In general/independent of custom elements, in "idomatic React" jsx props are in camelCase. These then can get predictably translated to attribute names by a simple `toLowerCase()`. There are also particular behaviors based on the type of value and the value itself.

With web components, things are a bit trickier since they cannot write generic predictable support under a wide variety of case (SSR, upgrades, props vs. attributes, etc. etc.). To account for all those permutations (including taking some lessons learned from weaknesses in Preact's built-in support for custom elements), they've settled on this table:

Has the custom element upgraded yet? | What is the type of the value? | Will we use attribute or property?
-- | -- | --
No (or SSR) | string/number | attribute
No (or SSR) | object/array/function | Do nothing
Yes | string/number | If ‘propname’ in element, Then use property.  Else use attribute
Yes | object/array/function | If ‘propname’ in element, Then use property. Else do nothing

where React properties of `propName` would be translated into `propname` in cases where they use attribute (and not property)

**Reply by heff** [reply]
Thanks for the summary. 🙏 

> These then can get predictably translated to attribute names by a simple toLowerCase()

That's great for JSX going from props to attrs, but not so helpful going the other way. With dashes it can work both ways, but I see that would be a breaking change for many existing attrs.

**Reply by luwes** [reply]
sorry, just saw this after I wrote the comment below. 

to me it looks like we don't need to remove the dashes at all, just need to ensure all elements have a corresponding camelCase property. or am I missing something?

with the knowledge that Media Chrome SSR is the one who renders the attributes like `media-preview-image` etc. and that  React idiomatic doesn't matter here

**Comment by cjpillsbury** [comment]
Also, per your other callout, I also agree that it should be called `media-preview-thumbnail` for consistency of `media-{FUNCTION}-{CONTROL_TYPE}` naming convention (e.g. `media-play-button`, `media-time-range`). So we don't surprisingly introduce a breaking change, we should probably consider registering both in the interim and causing a `console.warn` when folks use the old name. See, e.g.: https://github.com/muxinc/media-chrome/blob/main/src/js/index.js#L40-L53

**Reply by cjpillsbury** [reply]
Per our out of band convo, the deprecation warning shouldn't be necessary, since it is highly unlikely that many folks are directly using the `media-preview-thumbnail`

**Comment by luwes** [comment]
I didn't want to imply that big of a change actually 😄  but I do like the cleanliness of that approach 

to be clear initially I thought to pick those attributes that are less related to the media element, it makes sense to me that `media-current-time` and `media-volume` are called that because it's a value of the media element but some attributes are more specific to the custom element itself. in the case above `src` or `image` and then `coords`. those can be easily more generic. I admit it's a bit of a fuzzy line.

**Comment by heff** [comment]
### Preview naming

> media-thumbnail-preview -> media-preview-thumbnail

Yep makes sense to me. 👍 

### Dashes

I think I could be convinced to kill dashes, but they are pretty nice to have. They look nicer when attrs are name spaced, they make translating from same-case attrs to camel case props easier, and don't seem to be against [general guidelines](https://stackoverflow.com/questions/925994/what-characters-are-allowed-in-an-html-attribute-name#:~:text=Attribute%20names%20must%20consist%20of,are%20not%20defined%20by%20Unicode.).

> So, for custom element names you can use upper/lower alphanumeric, "_" underscore, ":" colon, or any of the unicode characters called out in the spec, as a start character, then dashes "-", dots ".", alpha etc... as body characters.

As far as mimicking HTML, I get why dashes look different. And if other projects like React are being short-sighted in what's allowed in attr names when handling them, that could create some friction.

### Name spacing attrs with "media"

This is one where I worry about some unintended consequences. Today "media-" attributes are set by media-controller (including media-preview-image and media-preview-coords). Some media chrome elements have attributes that are not "media-", and those are set by the developer. Without media-, I worry about having no clear way to distinguish between the two. 

And then secondly, potential collisions. Today we use observedAttributes to auto-detect which media attrs should be set on a custom element. Without a name space, we could end up setting an attribute that the custom element author was intending for something else. e.g. if we made media-preview-image -> src. We could change the use of observedAttributes but it's nice to not have to define the media attrs you care about in two places.

**Reply by cjpillsbury** [reply]
> And if other projects like React are being short-sighted in what's allowed in attr names when handling them, that could create some friction.

That's not exactly right. Attributes with hyphens are allowed, they just 

1. can't take advantage of the "property/attribute" relationship features described above (particularly relevant for consistency between "complex values" and "simple values", but with other benefits as well)
2. are harder to work with in javascript, since these would correspond to variable names or keys in objects
3. are not "idiomatic" to React and many other frameworks

**Reply by heff** [reply]
I think I'm convinced moving away from dashes for attrs. There's no issue with using camelCase in HTML attrs, right? At least in examples that could make longer attrs more readable, even if the casing is ignored.

**Reply by gkatsev** [reply]
HTML attributes are case-insensitive. You can even use `sPoNgEbOb cAsE` if you really felt like 😆.

**Reply by heff** [reply]
Yes. 😆

**Reply by cjpillsbury** [reply]
Yeah, as a specific callout of a divergence in the html spec vs. XML:

> All attribute names on [HTML elements](https://html.spec.whatwg.org/multipage/infrastructure.html#html-elements) in [HTML documents](https://dom.spec.whatwg.org/#html-document) get ASCII-lowercased automatically, so the restriction on ASCII uppercase letters doesn't affect such documents.

https://html.spec.whatwg.org/multipage/dom.html#custom-data-attribute

**Comment by heff** [comment]
The only other thing I've thought of for shortening attr names without creating potential attr name conflicts is allowing for some kind of mapping, like:

```html
<media-thumbnail-preview media-chrome-attributes="media-preview-image:src, media-preview-coords:coords">
```

I could see that being nice for UI elements that existed before or outside media chrome but can handle the attr values media chrome sets. That feels like probably a tiny amount of elements. And it feels like it sets up an ecosystem where we're regularly saying "over here it's called X, while over there it's called Y".

**Reply by heff** [reply]
With that, for the sake of coming to a decision, does anyone want to continue to make the case for shorter media attrs or a specific approach? My proposal would be to not shorten the attrs, but remove the dashes (and camelCase in all examples).

**Reply by luwes** [reply]
- I understand the media controller argument and to have a distinction from who set it. that makes sense, and it also makes it feel like less important to be concise in the naming, if we're not manually writing the HTML for it.

- using camcelCase attributes is very hard to get on board with for me. feels like a junior React dev for the first time wrote HTML or so. I've rarely seen this in plain HTML.

- removing the dashes I have yet to see any real proof that it's required for React 19 web components to work.
	I know from looking at the Preact source that they will attempt to set the property first so in JSX one would write the properties camelCase and they will be set. if you prefer to not target the props, just write lowercase attributes with dash or not, they just call setAttribute / removeAttribute under the hood.

the only thing we'd have to add I think is corresponding camelcase properties in each element.

link to Preact source setting props / attributes
https://github.com/preactjs/preact/blob/master/src/diff/props.js#L120-L143

**Reply by cjpillsbury** [reply]
If the camelCase props don't have corresponding attributes, the "fallback" cases for SSR and pre-upgrade will not work and will add incorrect attributes. See chart and responses to heff, above.
